core/src/verified_vote_packets.rs

```diff
@@ -158,12 +225,49 @@ impl VerifiedVotePackets {
                     let slot = vote.last_voted_slot().unwrap();
                     let hash = vote.hash();
 
-                    let validator_votes = self.0.entry(vote_account_key).or_default();
-                    validator_votes.insert((slot, hash), (packet_batch, signature));
-
-                    if validator_votes.len() > MAX_VOTES_PER_VALIDATOR {
-                        let smallest_key = validator_votes.keys().next().cloned().unwrap();
-                        validator_votes.remove(&smallest_key).unwrap();
+                    match (vote, is_full_tower_vote_enabled) {
+                        (VoteStateUpdate(_), true) => {
+                            let latest_slot_sent_to_bank = match self.0.get(&vote_account_key) {
+                                Some(vote) => vote.get_latest_slot_sent_to_bank(),
+                                _ => 0,
+                            };
+                            // Since votes are not incremental, we keep only the latest vote
+                            if slot > latest_slot_sent_to_bank {
+                                self.0.insert(
+                                    vote_account_key,
+                                    FullTowerVote(GossipVote {
+                                        slot,
+                                        hash,
+                                        packet_batch,
+                                        signature,
+                                    }),
+                                );
+                            }
+                        }
+                        _ => {
+                            let validator_votes: &mut BTreeMap<
+                                (Slot, Hash),
+                                (PacketBatch, Signature),
+                            > = match self
+                                .0
+                                .entry(vote_account_key)
+                                .or_insert(IncrementalVotes(BTreeMap::new()))
+                            {
+                                IncrementalVotes(votes) => votes,
+                                _ => {
+                                    error!(
+                                        "Originally {} submitted full tower votes, but now has reverted to incremental votes",
+                                        vote_account_key
+                                    );
+                                    continue;
+                                }
```

Comment:
Do you think we should continue supporting this case since we haven't completely removed the possibility of reverting/downgrading from the vote state update to the incremental vote state update. For instance, in the vote program we still support both.